### ISSUE TITLE
Plan for research ideas

### ISSUE THREAD

**Issue by eliird** [WestCoastML/VSLM#1]
# MultiDecode
- vlm beamsearch benchmark to compare with multiple decode
- reimplement the demo making performance optimization
- using multidecode for reasoning
- checkout the `Anthropics` multiple decode and summarize somewhere.

# Architecture
- Change the architecture in a way that the dimensions are decreasing with each transformer layer (kind of like CNN)
- Variation in sizes of layer of transformer (check `Brainstorming.md` architecture section)
- Layer 1 (token in concept out) -> Layer 2 (concept in concept out) -> Layer 3 (concept in token out) (kind of like BLT, take time to generate concepts)

# plan
- test diamond layer structure on GPT2 ,
- make an architecture that can do token to concept, concept to concept and concept to token
- test with a GPT2 size model.
- also look into style GAN kind of architecture would be possible for this

**Comment by eliird** [comment]
I read the gumble softmax paper the idea is very interesting. Its best to play with mnist to check if this works or not so I wrote some implementation of how sampling would work inside of neural network


```python
def gumbel_softmax(logits, temperature=1.0, hard=False):
    # Sample Gumbel noise,(maybe can be better)
    noise = -torch.log(-torch.log(torch.rand_like(logits) + 1e-10) + 1e-10)
    y = F.softmax((logits + noise) / temperature, dim=-1)
    
    if hard:
        # Convert to one-hot
        y_hard = torch.zeros_like(y)
        y_hard.scatter_(1, y.argmax(dim=1, keepdim=True), 1.0)
        y = (y_hard - y).detach() + y  # Straight-through estimator
    return y
```

so if you give it input logits os `sample_size * number_of_neurons` it will give a mask of which neurons should be sampled

![Image](https://github.com/user-attachments/assets/a49e9774-45cf-4e50-aad6-2039ed2f2eb7)

In the above image we sample 4 neurons from the 10 neurons, selecting 9th, 5th, 6th and 9th neuron from original layer. I am not sure if allowing repetition is a good idea or no.

**Comment by eliird** [comment]
Based on above, created a layer which can be used instead of linear layer

```python
class GumbelSamplingLayer(nn.Module):
    def __init__(self, in_dim, full_out_dim, sample_dim, temperature=1.0):
        super().__init__()
        self.linear = nn.Linear(in_dim, full_out_dim)
        self.logits = nn.Parameter(torch.randn(sample_dim, full_out_dim))  # learnable selector
        self.sample_dim = sample_dim
        self.temperature = temperature

    def forward(self, x):
        # x: (batch_size, in_dim)
        full_out = self.linear(x)  # (batch_size, full_out_dim)

        # Sample selection mask: (sample_dim, full_out_dim)
        mask_weights = gumbel_softmax(self.logits, temperature=self.temperature, hard=True)

        # Weighted sum: project from full_out_dim -> sample_dim
        # Output shape: (batch_size, sample_dim)
        sampled_out = torch.matmul(mask_weights, full_out.T).T
        return sampled_out
```

and using created neural network like below

```python
# Model using GumbelSamplingLayer
class GumbelNet(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc1 = GumbelSamplingLayer(784, 512, 256)
        self.relu = nn.ReLU()
        self.fc2 = GumbelSamplingLayer(256, 256, 128)
        self.output = nn.Linear(128, 10)

    def forward(self, x):
        x = x.view(-1, 28 * 28)
        x = self.relu(self.fc1(x))
        x = self.relu(self.fc2(x))
        return self.output(x)
############################################
class SimpleNet(nn.Module):
    def __init__(self):
        super().__init__()
        self.model = nn.Sequential(
            nn.Linear(784, 256),
            nn.ReLU(),
            nn.Linear(256, 128),
            nn.ReLU(),
            nn.Linear(128, 10),
        )

    def forward(self, x):
        x = x.view(-1, 28 * 28)
        return self.model(x)
```
I trained it for 25 epochs to train the above neural networks on mnist dataset and the results are as follows.

**GumbelNet**

![Image](https://github.com/user-attachments/assets/fd7b90db-22a8-4010-bba1-8615b4e93f80)

**SimpleNet**

![Image](https://github.com/user-attachments/assets/f0a3ac1b-4e60-4b6f-b4ab-67a0c4866789)

The convergance using gumble layers is slower compare to the normal layers. I will run it for twice the time to see if it will get to the same point or no.

**Comment by eliird** [comment]
I appended training loss two times so the plot is longer but it is for 25 epochs

**Comment by eliird** [comment]
![Image](https://github.com/user-attachments/assets/602ca682-909a-449c-a248-9b2143c335af)

Running for twice the epochs made the model learn more (still lower than the normal mnist)

**Comment by eliird** [comment]
## using this idea in LLMs for a baseline (writing random thoughts in my head)

-  We talked last time about networks being diamond `<>` shape last time but `><` hourglass (encoder, decoder bottleneck) is another possibility.

- so for every transformer layer we decrease the num of neurons sampled from MLP output and feed those to next transformer layer (which in a way is forcing each layer to look at less number of features).

**Comment by eliird** [comment]
created a PR for code used to generate the above graphs

**Comment by rstager** [comment]
Did you try to train it with hard = True
